### PR TITLE
fix(federation): fixed a bug where cached node costs were not reset after modification

### PR DESCRIPTION
**Pull request by duckki**
Fetch dependency graph nodes cache their cost values. And whenever their selection sets get modified, their cached cost must reset. However, we had a few places where that invariant was violated.

The `selection_set_mut` method is supposed to be used, since the method resets the cache. However, other methods could get around it and modify `selection_set` without calling the method. This PR fixes those places to use the `selection_set_mut` method.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

[ROUTER-553]: https://apollographql.atlassian.net/browse/ROUTER-553?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs [hunk 1]
```diff
@@ -2361,9 +2361,9 @@ impl FetchDependencyGraphNode {
     }
 
     fn remove_inputs_from_selection(&mut self) -> Result<(), FederationError> {
-        let fetch_selection_set = &mut self.selection_set;
         if let Some(inputs) = &mut self.inputs {
             self.cached_cost = None;
+            let fetch_selection_set = &mut self.selection_set;
```
Comment (duckki) [inline apollo-federation/src/query_plan/fetch_dependency_graph.rs:2366]: Is is an unrelated refactoring.

**Review by dariuszkuc — APPROVED** [review]
